src/lib/operators/materialize.cpp

```diff
@@ -0,0 +1,92 @@
+#include "materialize.hpp"
+
+#include <mutex>
+
+#include "resolve_type.hpp"
+#include "storage/base_column.hpp"
+#include "storage/chunk.hpp"
+#include "storage/create_iterable_from_column.hpp"
+#include "storage/table.hpp"
+#include "scheduler/current_scheduler.hpp"
+#include "scheduler/job_task.hpp"
+
+namespace opossum {
+
+Materialize::Materialize(const std::shared_ptr<const AbstractOperator> in)
+    : AbstractReadOnlyOperator{OperatorType::Materialize, in} {}
+
+const std::string Materialize::name() const { return "Materialize"; }
+const std::string Materialize::description(DescriptionMode description_mode) const { return name(); }
+
+std::shared_ptr<AbstractOperator> Materialize::_on_recreate(
+    const std::vector<AllParameterVariant>& args, const std::shared_ptr<AbstractOperator>& recreated_input_left,
+    const std::shared_ptr<AbstractOperator>& recreated_input_right) const {
+  return std::make_shared<Materialize>(recreated_input_left);
+}
+
+std::shared_ptr<const Table> Materialize::_on_execute() {
+  const auto in_table = _input_left->get_output();
+
+  auto out_table = std::make_shared<Table>(in_table->column_definitions(), TableType::Data);
+
+  std::mutex output_mutex{};
+
+  auto jobs = std::vector<std::shared_ptr<AbstractTask>>{};
+  jobs.reserve(in_table->chunk_count());
+
+  for (ChunkID chunk_id{0u}; chunk_id < in_table->chunk_count(); ++chunk_id) {
+    auto job_task = std::make_shared<JobTask>([=, &output_mutex]() {
+      const auto in_chunk = in_table->get_chunk(chunk_id);
+
+      // The new chunk is allocated on the same node
+      const auto& alloc = in_chunk->get_allocator();
+
+      ChunkColumns out_columns{alloc};
+      out_columns.reserve(in_chunk->column_count());
+
+      for (ColumnID column_id{0u}; column_id < in_chunk->column_count(); ++column_id) {
+        const auto in_base_column = in_chunk->get_mutable_column(column_id);
+        const auto data_type = in_table->column_data_type(column_id);
+
+        if (std::dynamic_pointer_cast<BaseValueColumn>(in_base_column) != nullptr) {
```

Comment:
This could use a comment - “If the input column already is a value column, we don’t have to do anything and pass it forward.”

Reply:
It is a fundamental property of Tables that they are either Data or References. With this knowledge in mind it is somewhat confusing (albeit more flexible towards potential changes of the invariant) to see the handling of this fundamental property at Column level. 
You could just detect Data tables at the top of the operator and then forward the entire table.

Reply:
#782? ;)

Reply:
@mrzzzrm have a second look. I’m only checking if the columns is a **value** column.